File: SOURCE_NATIVE/EW4_iOS_NativePort/NativeCore/Sources/EW4NativeCore/CountryTurnCore.swift
import Foundation

public struct CountryResources: Equatable, Codable, Sendable {
    public var money: Int
    public var industry: Int
    public var food: Int

    public init(money: Int, industry: Int, food: Int) {
        self.money = money
        self.industry = industry
        self.food = food
    }
}

public enum BattleMode: String, Sendable {
    case campaign
    case conquest
    case tutorial
}

public enum CountryRelation: String, Sendable {
    case ally
    case hostile
    case neutral
}

public enum CountryTurnCore {
    public static let sentinel: UInt64 = 0xffff_ffff

    public static func cleanResource(_ value: UInt64?, fallback: Int = 0) -> Int {
        guard let value,
              value != sentinel,
              value <= 10_000_000 else {
            return max(0, fallback)
        }
        return Int(value)
    }

    public static func headerResources(_ battle: BattleRecord) -> CountryResources {
        let raw = battle.header.raw ?? []
        func value(at index: Int) -> UInt64? {
            index < raw.count ? raw[index] : nil
        }

        return CountryResources(
            money: cleanResource(value(at: 18), fallback: 800),
            industry: cleanResource(value(at: 19), fallback: 300),
            food: cleanResource(value(at: 20), fallback: 500)
        )
    }

    public static func countryResources(
        _ country: BattleCountry,
        fallback: CountryResources = CountryResources(money: 0, industry: 0, food: 0)
    ) -> CountryResources {
        let raw = country.rawResources ?? []
        func value(at index: Int) -> UInt64? {
            index < raw.count ? raw[index] : nil
        }

        return CountryResources(
            money: cleanResource(value(at: 30), fallback: fallback.money),
            industry: cleanResource(value(at: 31), fallback: fallback.industry),
            food: cleanResource(value(at: 32), fallback: fallback.food)
        )
    }

    public static func buildLedgers(
        _ battle: BattleRecord,
        mode: BattleMode = .campaign,
        playerOwner: Int = 0,
        restored: [Int: CountryResources]? = nil
    ) -> [Int: CountryResources] {
        if let restored {
            return restored.mapValues {
                CountryResources(
                    money: max(0, $0.money),
                    industry: max(0, $0.industry),
                    food: max(0, $0.food)
                )
            }
        }

        let header = headerResources(battle)
        var ledgers: [Int: CountryResources] = [:]

        for country in battle.countries {
            ledgers[country.index] = countryResources(
                country,
                fallback: CountryResources(money: 0, industry: 0, food: header.food)
            )
        }

        if mode != .conquest {
            ledgers[playerOwner] = header
        }

        return ledgers
    }

    public static func hint(_ battle: BattleRecord, owner: Int?) -> Int? {
        guard let owner, owner != 255 else {
            return 4
        }

        if let group = battle.relationGroups?[String(owner)] {
            return group
        }

        return battle.countries.first(where: { $0.index == owner })?.relationHint
    }

    public static func relation(
        _ battle: BattleRecord,
        _ a: Int,
        _ b: Int,
        mode: BattleMode = .campaign,
        playerOwner: Int = 0
    ) -> CountryRelation {
        if a == b {
            return .ally
        }

        if a == 255 || b == 255 {
            return .neutral
        }

        if mode != .conquest {
            return (a == playerOwner || b == playerOwner) ? .hostile : .ally
        }

        let aHint = hint(battle, owner: a)
        let bHint = hint(battle, owner: b)

        if aHint == 4 || bHint == 4 {
            return .neutral
        }

        if let aHint,
           let bHint,
           (aHint == 1 || aHint == 2),
           (bHint == 1 || bHint == 2) {
            return aHint == bHint ? .ally : .hostile
        }

        return .hostile
    }

    public static func aiTurnOrder(
        _ battle: BattleRecord,
        livingOwners: Set<Int>,
        playerOwner: Int,
        mode: BattleMode = .campaign
    ) -> [Int] {
        var order: [Int] = []

        // Preserve the authored country order from the BTL record first.
        for country in battle.countries {
            let owner = country.index
            if owner == playerOwner || !livingOwners.contains(owner) {
                continue
            }
            if mode == .conquest && hint(battle, owner: owner) == 4 {
                continue
            }
            order.append(owner)
        }

        // Keep any live owner omitted from the country table deterministic.
        for owner in livingOwners.sorted() {
            if owner == playerOwner || order.contains(owner) {
                continue
            }
            if mode == .conquest && hint(battle, owner: owner) == 4 {
                continue
            }
            order.append(owner)
        }

        return order
    }
}

File: SOURCE_LEAN/EW4_Web_Port_v0.61/country_turn_core.js
'use strict';
(function(root,factory){
  const api=factory();
  if(typeof module==='object'&&module.exports)module.exports=api;
  root.EW4CountryTurn=api;
})(typeof globalThis!=='undefined'?globalThis:this,function(){
  const SENTINEL=0xffffffff;
  function cleanResource(v,fallback=0){
    v=Number(v);
    if(!Number.isFinite(v)||v<0||v===SENTINEL||v>10000000)return Math.max(0,Number(fallback)||0);
    return Math.floor(v);
  }
  function headerResources(battle){
    const raw=battle?.header?.raw||[];
    return{money:cleanResource(raw[18],800),industry:cleanResource(raw[19],300),food:cleanResource(raw[20],500)};
  }
  function countryResources(country,fallback={money:0,industry:0,food:0}){
    const raw=country?.raw_u32_36_180||[];
    // High-confidence BTL country economy triplet. Keep this decoder isolated until native names are confirmed.
    return{
      money:cleanResource(raw[30],fallback.money),
      industry:cleanResource(raw[31],fallback.industry),
      food:cleanResource(raw[32],fallback.food)
    };
  }
  function buildLedgers(battle,{mode='campaign',playerOwner=0,restored=null}={}){
    if(restored&&typeof restored==='object'){
      const out={};
      for(const [k,v] of Object.entries(restored))out[String(+k)]={money:cleanResource(v?.money),industry:cleanResource(v?.industry),food:cleanResource(v?.food)};
      return out;
    }
    const h=headerResources(battle),out={};
    for(const c of battle?.countries||[])out[String(c.index)]=countryResources(c,{money:0,industry:0,food:h.food});
    // Campaign/tutorial headers carry the stage-specific player treasury used by the current port.
    // Conquest instead uses each selected country's own BTL country record.
    if(mode!=='conquest')out[String(+playerOwner)]={...h};
    return out;
  }
  function hint(battle,owner){
    if(owner==null||+owner===255)return 4;
    const g=battle?.relation_groups?.[String(+owner)];
    if(g!==undefined&&g!==null)return +g;
    const c=battle?.countries?.[+owner];
    return c?.relation_hint==null?null:+c.relation_hint;
  }
  function relation(battle,a,b,{mode='campaign',playerOwner=0}={}){
    a=+a;b=+b;playerOwner=+playerOwner;
    if(a===b)return'ally';
    if(a===255||b===255)return'neutral';
    if(mode!=='conquest'){
      // Campaign relation_hint is NOT a conquest alliance group. Preserve the recovered 0.61
      // player-centric hostility model while preventing separate enemy countries from fighting each other.
      if(a===playerOwner||b===playerOwner)return'hostile';
      return'ally';
    }
    const ha=hint(battle,a),hb=hint(battle,b);
    if(ha===4||hb===4)return'neutral';
    if((ha===1||ha===2)&&(hb===1||hb===2))return ha===hb?'ally':'hostile';
    // Some decoded final conquest country records still have an unresolved relation byte. Preserve old runtime behavior conservatively.
    return'hostile';
  }
  function livingOwners(units){return new Set((units||[]).filter(u=>!u.dead).map(u=>+u.owner).filter(x=>x!==255));}
  function aiTurnOrder(battle,units,playerOwner,mode='campaign'){
    const live=livingOwners(units),out=[];
    for(const c of battle?.countries||[]){
      const owner=+c.index;if(owner===+playerOwner||!live.has(owner))continue;
      if(mode==='conquest'&&hint(battle,owner)===4)continue;
      out.push(owner);
    }
    // Preserve any decoded owner not represented in countries[] rather than dropping its units.
    for(const owner of live)if(owner!==+playerOwner&&!out.includes(owner)&&(mode!=='conquest'||hint(battle,owner)!==4))out.push(owner);
    return out;
  }
  return{cleanResource,headerResources,countryResources,buildLedgers,hint,relation,livingOwners,aiTurnOrder};
});
